From e3108b034923ab6dea1266106bb2260251aa4acf Mon Sep 17 00:00:00 2001
From: Louis Marelle <louis.marelle@latmos.ipsl.fr>
Date: Tue, 9 Oct 2018 15:12:00 +0200
Subject: [PATCH 06/52] Add a BB emission package for SAPRC99

Adds a biomass burning (BB) emission package for SAPRC99 mechanisms,
BIOMASSB_SAPRC (biomass_burn_opt = 6 in &chem), and couples it to the
plume rise model and to MOSAIC aerosols. If this BIOMASSB_SAPRC option
is selected, ebu(p_ebu_oc) is added to VBS2 species in chem in
chem/emissions_driver.F, and anthropogenic emissions can be either added
through emiss_ant(p_e_orgj_a) or emiss_ant(p_e_orgj). If BIOMASSB_SAPRC
is not used, emissions_driver defaults to the old behavior, and organic
aerosols are added to VBS2 species through wrfchemi files only, e.g.
through emiss_ant(p_e_orgj_a) and emiss_ant(p_e_orgj_bb)
---
 Registry/registry.chem        | 47 ++++++++++++++++++++
 chem/emissions_driver.F       | 83 +++++++++++++++++++++++++----------
 chem/module_add_emiss_burn.F  | 50 ++++++++++++++++++++-
 chem/module_mosaic_addemiss.F |  7 +--
 chem/module_plumerise1.F      | 45 +++++++++++++++++++
 5 files changed, 205 insertions(+), 27 deletions(-)

diff --git a/Registry/registry.chem b/Registry/registry.chem
index 70586eae..cf4142d3 100644
--- a/Registry/registry.chem
+++ b/Registry/registry.chem
@@ -413,6 +413,28 @@ state    real  ebu_hcooh    ikjf     ebu           1         Z        -      "eb
 state    real  ebu_c2h2     ikjf     ebu           1         Z        -      "ebu_c2h2"      "biomass burning emiss"  "mol km^-2 hr^-1"
 state    real  ebu_hpald    ikjf     ebu           1         Z        -      "ebu_hpald"     "biomass burning emiss"  "mol km^-2 hr^-1"
 state    real  ebu_xylenes  ikjf     ebu           1         Z        -      "ebu_xylenes"   "biomass burning emiss"  "mol km^-2 hr^-1"
+# additional arrays for SAPRC99 biomass burning
+state    real  ebu_acet     ikjf     ebu           1         Z        -      "ebu_acet"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_alk3     ikjf     ebu           1         Z        -      "ebu_alk3"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_alk4     ikjf     ebu           1         Z        -      "ebu_alk4"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_alk5     ikjf     ebu           1         Z        -      "ebu_alk5"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_aro1     ikjf     ebu           1         Z        -      "ebu_aro1"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_aro2     ikjf     ebu           1         Z        -      "ebu_aro2"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_bald     ikjf     ebu           1         Z        -      "ebu_bald"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_ccho     ikjf     ebu           1         Z        -      "ebu_ccho"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_cco_oh   ikjf     ebu           1         Z        -      "ebu_cco_oh"    "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_ethene   ikjf     ebu           1         Z        -      "ebu_ethene"    "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_hono     ikjf     ebu           1         Z        -      "ebu_hono"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_isoprene ikjf     ebu           1         Z        -      "ebu_isoprene"  "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_meoh     ikjf     ebu           1         Z        -      "ebu_meoh"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_methacro ikjf     ebu           1         Z        -      "ebu_methacro"  "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_ole1     ikjf     ebu           1         Z        -      "ebu_ole1"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_ole2     ikjf     ebu           1         Z        -      "ebu_ole2"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_phen     ikjf     ebu           1         Z        -      "ebu_phen"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_prod2    ikjf     ebu           1         Z        -      "ebu_prod2"     "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_rcho     ikjf     ebu           1         Z        -      "ebu_rcho"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_rno3     ikjf     ebu           1         Z        -      "ebu_rno3"      "biomass burning emiss"  "mol km^-2 hr^-1"
+state    real  ebu_terp     ikjf     ebu           1         Z        -      "ebu_terp"      "biomass burning emiss"  "mol km^-2 hr^-1"
 
 # CH4 biomass burning emissions
 state    real  ebu_ch4      ikjf     ebu           1         Z        h      "ebu_ch4"       "biomass burning emiss"  "mol km^-2 hr^-1"
@@ -485,6 +507,28 @@ state    real  ebu_in_hcooh       i]jf      ebu_in      1         -        i07
 state    real  ebu_in_c2h2        i]jf      ebu_in      1         -        i07        "ebu_in_c2h2"       "EMISSIONS"          "mol km^-2 hr^-1"
 state    real  ebu_in_hpald       i]jf      ebu_in      1         -        i07        "ebu_in_hpald"      "EMISSIONS"          "mol km^-2 hr^-1"
 state    real  ebu_in_xylenes     i]jf      ebu_in      1         -        i07        "ebu_in_xylenes"    "EMISSIONS"          "mol km^-2 hr^-1"
+# additional arrays for SAPRC99 biomass burning
+state    real  ebu_in_acet        iljf     ebu_in       1         -        i07        "ebu_in_acet"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_alk3        iljf     ebu_in       1         -        i07        "ebu_in_alk3"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_alk4        iljf     ebu_in       1         -        i07        "ebu_in_alk4"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_alk5        iljf     ebu_in       1         -        i07        "ebu_in_alk5"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_aro1        iljf     ebu_in       1         -        i07        "ebu_in_aro1"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_aro2        iljf     ebu_in       1         -        i07        "ebu_in_aro2"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_bald        iljf     ebu_in       1         -        i07        "ebu_in_bald"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_ccho        iljf     ebu_in       1         -        i07        "ebu_in_ccho"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_cco_oh      iljf     ebu_in       1         -        i07        "ebu_in_cco_oh"     "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_ethene      iljf     ebu_in       1         -        i07        "ebu_in_ethene"     "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_hono        iljf     ebu_in       1         -        i07        "ebu_in_hono"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_isoprene    iljf     ebu_in       1         -        i07        "ebu_in_isoprene"   "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_meoh        iljf     ebu_in       1         -        i07        "ebu_in_meoh"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_methacro    iljf     ebu_in       1         -        i07        "ebu_in_methacro"   "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_ole1        iljf     ebu_in       1         -        i07        "ebu_in_ole1"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_ole2        iljf     ebu_in       1         -        i07        "ebu_in_ole2"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_phen        iljf     ebu_in       1         -        i07        "ebu_in_phen"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_prod2       iljf     ebu_in       1         -        i07        "ebu_in_prod2"      "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_rcho        iljf     ebu_in       1         -        i07        "ebu_in_rcho"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_rno3        iljf     ebu_in       1         -        i07        "ebu_in_rno3"       "EMISSIONS"          "mol km^-2 hr^-1"
+state    real  ebu_in_terp        iljf     ebu_in       1         -        i07        "ebu_in_terp"       "EMISSIONS"          "mol km^-2 hr^-1"
 
 # additional array for ch4 bbm emission
 state    real  ebu_in_ch4         i]jf      ebu_in      1         -        i07        "ebu_in_ch4"        "EMISSIONS"          "mol km^-2 hr^-1"
@@ -4098,6 +4142,9 @@ package   biomassb_t1_mozcart   biomass_burn_opt==4          -          state:me
 # GHG biomass burning
 package   biomassb_ghg  biomass_burn_opt==5          -             state::mean_fct_agtf,mean_fct_agef,mean_fct_agsv,mean_fct_aggr,firesize_agtf,firesize_agef,firesize_agsv,firesize_aggr;ebu:ebu_co,ebu_co2,ebu_ch4;ebu_in:ebu_in_co,ebu_in_co2,ebu_in_ch4
 
+# SAPRC99 biomass burning package
+package   biomassb_saprc       biomass_burn_opt==6             -          state:mean_fct_agtf,mean_fct_agef,mean_fct_agsv,mean_fct_aggr,firesize_agtf,firesize_agef,firesize_agsv,firesize_aggr;ebu:ebu_co,ebu_no,ebu_no2,ebu_so2,ebu_nh3,ebu_ch4,ebu_acet,ebu_c2h6,ebu_c3h8,ebu_alk3,ebu_alk4,ebu_alk5,ebu_aro1,ebu_aro2,ebu_bald,ebu_ccho,ebu_cco_oh,ebu_ethene,ebu_hcho,ebu_hcooh,ebu_hono,ebu_isoprene,ebu_mek,ebu_meoh,ebu_methacro,ebu_mgly,ebu_mvk,ebu_ole1,ebu_ole2,ebu_phen,ebu_prod2,ebu_rcho,ebu_rno3,ebu_terp,ebu_oc,ebu_bc,ebu_sulf;ebu_in:ebu_in_co,ebu_in_no,ebu_in_no2,ebu_in_so2,ebu_in_nh3,ebu_in_ch4,ebu_in_acet,ebu_in_c2h6,ebu_in_c3h8,ebu_in_alk3,ebu_in_alk4,ebu_in_alk5,ebu_in_aro1,ebu_in_aro2,ebu_in_bald,ebu_in_ccho,ebu_in_cco_oh,ebu_in_ethene,ebu_in_hcho,ebu_in_hcooh,ebu_in_hono,ebu_in_isoprene,ebu_in_mek,ebu_in_meoh,ebu_in_methacro,ebu_in_mgly,ebu_in_mvk,ebu_in_ole1,ebu_in_ole2,ebu_in_phen,ebu_in_prod2,ebu_in_rcho,ebu_in_rno3,ebu_in_terp,ebu_in_oc,ebu_in_bc,ebu_in_sulf 
+
 # Package for mozart pH diagnostics
 package  do_mozart_ph_diag    mozart_ph_diag==1      -             state:ph_cw,ph_aer01,ph_aer02,ph_aer03,ph_aer04
 
diff --git a/chem/emissions_driver.F b/chem/emissions_driver.F
index 9c4c8cb1..2222b74e 100644
--- a/chem/emissions_driver.F
+++ b/chem/emissions_driver.F
@@ -631,6 +631,7 @@ CONTAINS
       IF ( config_flags%biomass_burn_opt == BIOMASSB_MOZC .OR. &
            config_flags%biomass_burn_opt == BIOMASSB_T1_MOZCART  .OR. &
            config_flags%biomass_burn_opt == BIOMASSB_MOZ  .OR. &
+           config_flags%biomass_burn_opt == BIOMASSB_SAPRC .OR. &
            config_flags%biomass_burn_opt == BIOMASSB_GHG  .OR. &
            config_flags%biomass_burn_opt == BIOMASSB ) then
         IF ( ktau==1 ) then
@@ -784,7 +785,7 @@ CONTAINS
 !!! ***********   FIRE AND CHEM ****************************************
 !
     fire_select:  SELECT CASE(config_flags%biomass_burn_opt)
-     CASE (BIOMASSB,BIOMASSB_MOZC,BIOMASSB_MOZ,BIOMASSB_T1_MOZCART,BIOMASSB_GHG)
+     CASE (BIOMASSB,BIOMASSB_MOZC,BIOMASSB_MOZ,BIOMASSB_T1_MOZCART,BIOMASSB_GHG,BIOMASSB_SAPRC)
 !      if( do_plumerisefire )then
 !         CALL wrf_debug(15,'fire emissions: calling biomassb')
 !         write(0,*)ktau,stepfirepl
@@ -1340,31 +1341,69 @@ CONTAINS
                 chem(i,k,j,p_nh3)  = chem(i,k,j,p_nh3)                       &
                      +emis_ant(i,k,j,p_e_nh3)*conv
 
-!Use OM/OC of 1.25 for fossil and OM:OC of 1.57 for biomass
-! O:C=0.06,H:C=1.8, N:C=0.02 for fossil : OM/OC=(16*0.06+12+14*0.02+12)/12=1.25
-!O:C=0.3 H:C=1.8, N:C=0.02 for biomass, OM/OC=1.57 for biomass
-
-         chem(i,k,j,p_pcg1_b_c)  =  chem(i,k,j,p_pcg1_b_c)                        &
-        +(emis_ant(i,k,j,p_e_orgi_bb)/1.57+emis_ant(i,k,j,p_e_orgj_bb)/1.57)*conv3*1.17
-        chem(i,k,j,p_pcg2_b_c)  =  chem(i,k,j,p_pcg2_b_c)                        &
-        +(emis_ant(i,k,j,p_e_orgi_bb)/1.57+emis_ant(i,k,j,p_e_orgj_bb)/1.57)*conv3*7.605
-        chem(i,k,j,p_pcg1_f_c)  =  chem(i,k,j,p_pcg1_f_c)                        &
-        +(emis_ant(i,k,j,p_e_orgi_a)/1.25+emis_ant(i,k,j,p_e_orgj_a)/1.25)*conv3*1.17
-        chem(i,k,j,p_pcg2_f_c)  =  chem(i,k,j,p_pcg2_f_c)                        &
-        +(emis_ant(i,k,j,p_e_orgi_a)/1.25+emis_ant(i,k,j,p_e_orgj_a)/1.25)*conv3*7.605
-        chem(i,k,j,p_pcg1_b_o)  =  chem(i,k,j,p_pcg1_b_o)                        &
-        +(emis_ant(i,k,j,p_e_orgi_bb)/1.57+emis_ant(i,k,j,p_e_orgj_bb)/1.57)*conv3*0.40
-        chem(i,k,j,p_pcg2_b_o)  =  chem(i,k,j,p_pcg2_b_o)                        &
-        +(emis_ant(i,k,j,p_e_orgi_bb)/1.57+emis_ant(i,k,j,p_e_orgj_bb)/1.57)*conv3*2.60
-        chem(i,k,j,p_pcg1_f_o)  =  chem(i,k,j,p_pcg1_f_o)                        &
-        +(emis_ant(i,k,j,p_e_orgi_a)/1.25+emis_ant(i,k,j,p_e_orgj_a)/1.25)*conv3*0.08
-        chem(i,k,j,p_pcg2_f_o)  =  chem(i,k,j,p_pcg2_f_o)                        &
-        +(emis_ant(i,k,j,p_e_orgi_a)/1.25+emis_ant(i,k,j,p_e_orgj_a)/1.25)*conv3*0.52
-
+                ! If the SAPRC biomass burning package is not used, add BB
+                ! organic aerosol emissions as e.g. emis_ant(i,k,j,p_e_orgj_bb),
+                ! instead of as ebu array instead
+                if ( config_flags%biomass_burn_opt /= BIOMASSB_SAPRC ) then
+                  !Use OM/OC of 1.25 for fossil and OM:OC of 1.57 for biomass
+                  ! O:C=0.06,H:C=1.8, N:C=0.02 for fossil : OM/OC=(16*0.06+12+14*0.02+12)/12=1.25
+                  !O:C=0.3 H:C=1.8, N:C=0.02 for biomass, OM/OC=1.57 for biomass
+                   chem(i,k,j,p_pcg1_b_c)  =  chem(i,k,j,p_pcg1_b_c)                        &
+                  +(emis_ant(i,k,j,p_e_orgi_bb)/1.57+emis_ant(i,k,j,p_e_orgj_bb)/1.57)*conv3*1.17
+                  chem(i,k,j,p_pcg2_b_c)  =  chem(i,k,j,p_pcg2_b_c)                        &
+                  +(emis_ant(i,k,j,p_e_orgi_bb)/1.57+emis_ant(i,k,j,p_e_orgj_bb)/1.57)*conv3*7.605
+                  chem(i,k,j,p_pcg1_f_c)  =  chem(i,k,j,p_pcg1_f_c)                        &
+                  +(emis_ant(i,k,j,p_e_orgi_a)/1.25+emis_ant(i,k,j,p_e_orgj_a)/1.25)*conv3*1.17
+                  chem(i,k,j,p_pcg2_f_c)  =  chem(i,k,j,p_pcg2_f_c)                        &
+                  +(emis_ant(i,k,j,p_e_orgi_a)/1.25+emis_ant(i,k,j,p_e_orgj_a)/1.25)*conv3*7.605
+                  chem(i,k,j,p_pcg1_b_o)  =  chem(i,k,j,p_pcg1_b_o)                        &
+                  +(emis_ant(i,k,j,p_e_orgi_bb)/1.57+emis_ant(i,k,j,p_e_orgj_bb)/1.57)*conv3*0.40
+                  chem(i,k,j,p_pcg2_b_o)  =  chem(i,k,j,p_pcg2_b_o)                        &
+                  +(emis_ant(i,k,j,p_e_orgi_bb)/1.57+emis_ant(i,k,j,p_e_orgj_bb)/1.57)*conv3*2.60
+                  chem(i,k,j,p_pcg1_f_o)  =  chem(i,k,j,p_pcg1_f_o)                        &
+                  +(emis_ant(i,k,j,p_e_orgi_a)/1.25+emis_ant(i,k,j,p_e_orgj_a)/1.25)*conv3*0.08
+                  chem(i,k,j,p_pcg2_f_o)  =  chem(i,k,j,p_pcg2_f_o)                        &
+                  +(emis_ant(i,k,j,p_e_orgi_a)/1.25+emis_ant(i,k,j,p_e_orgj_a)/1.25)*conv3*0.52
+
+                end if ! if ( config_flags%biomass_burn_opt /= BIOMASSB_SAPRC )
 
              end do
           end do
        end do
+       ! If the SAPRC biomass burning package is used, add BB organic aerosol
+       ! emissions as ebu(i,k,j,p_ebu_oc) instead of e.g.
+       ! emis_ant(i,k,j,p_e_orgj_bb), and allow the use of e_orgj instead of
+       ! e_orgj_a for anthropogenic emissions
+       if ( config_flags%biomass_burn_opt == BIOMASSB_SAPRC ) then
+         do j=jts,jte
+           do i=its,ite
+             do k=kts,kte
+               ! Use OM/OC of 1.25 for fossil and OM:OC of 1.57 for biomass
+               ! O:C=0.06,H:C=1.8, N:C=0.02 for fossil : OM/OC=(16*0.06+12+14*0.02+12)/12=1.25
+               ! O:C=0.3 H:C=1.8, N:C=0.02 for biomass, OM/OC=1.57 for biomass
+               conv3 = (dtstep/dz8w(i,k,j))*alt(i,k,j)*28/250*1e-3! Molecular weight of C(15)H(27)N(0.3)O(4.5)
+
+               chem(i,k,j,p_pcg1_b_c)  =  chem(i,k,j,p_pcg1_b_c) + (ebu(i,k,j,p_ebu_oc) / 1.57) * conv3 * 1.17
+               chem(i,k,j,p_pcg2_b_c)  =  chem(i,k,j,p_pcg2_b_c) + (ebu(i,k,j,p_ebu_oc) / 1.57) * conv3 * 7.605
+               chem(i,k,j,p_pcg1_f_c)  =  chem(i,k,j,p_pcg1_f_c)  &
+                 + (emis_ant(i,k,j,p_e_orgi) / 1.25 + emis_ant(i,k,j,p_e_orgi_a) / 1.25 &
+                 + emis_ant(i,k,j,p_e_orgj) / 1.25 + emis_ant(i,k,j,p_e_orgj_a) / 1.25) * conv3 * 1.17
+               chem(i,k,j,p_pcg2_f_c)  =  chem(i,k,j,p_pcg2_f_c) &
+                 + (emis_ant(i,k,j,p_e_orgi) / 1.25 + emis_ant(i,k,j,p_e_orgi_a) / 1.25 &
+                 + emis_ant(i,k,j,p_e_orgj) / 1.25 + emis_ant(i,k,j,p_e_orgj_a) / 1.25) * conv3 * 7.605
+               chem(i,k,j,p_pcg1_b_o)  =  chem(i,k,j,p_pcg1_b_o) + (ebu(i,k,j,p_ebu_oc) / 1.57) * conv3 * 0.40
+               chem(i,k,j,p_pcg2_b_o)  =  chem(i,k,j,p_pcg2_b_o) + (ebu(i,k,j,p_ebu_oc) / 1.57) * conv3 * 2.60
+               chem(i,k,j,p_pcg1_f_o)  =  chem(i,k,j,p_pcg1_f_o) &
+                 + (emis_ant(i,k,j,p_e_orgi) / 1.25 + emis_ant(i,k,j,p_e_orgi_a) / 1.25 &
+                 + emis_ant(i,k,j,p_e_orgj) / 1.25 + emis_ant(i,k,j,p_e_orgj_a) / 1.25) * conv3 * 0.08
+               chem(i,k,j,p_pcg2_f_o)  =  chem(i,k,j,p_pcg2_f_o) &
+                 + (emis_ant(i,k,j,p_e_orgi) / 1.25 + emis_ant(i,k,j,p_e_orgi_a) / 1.25 &
+                 + emis_ant(i,k,j,p_e_orgj) / 1.25 + emis_ant(i,k,j,p_e_orgj_a) / 1.25) * conv3 * 0.52
+
+             end do ! do j=jts,jte
+           end do ! do i=its,ite
+         end do ! do k=kts,kte
+       end if ! if ( config_flags%biomass_burn_opt == BIOMASSB_SAPRC ) then
       else
        do j=jts,jte
           do i=its,ite
diff --git a/chem/module_add_emiss_burn.F b/chem/module_add_emiss_burn.F
index 76a33a80..722fc51f 100644
--- a/chem/module_add_emiss_burn.F
+++ b/chem/module_add_emiss_burn.F
@@ -294,8 +294,7 @@ CONTAINS
         enddo
         enddo
         enddo
-      CASE (RADM2,RACM_KPP,RACM_MIM_KPP,SAPRC99_KPP,SAPRC99_MOSAIC_4BIN_VBS2_KPP, &
-           SAPRC99_MOSAIC_8BIN_VBS2_AQ_KPP, SAPRC99_MOSAIC_8BIN_VBS2_KPP)!BSINGH(12/03/2013): Added SAPRC 8 bin and non-aq on (04/07/2014)
+      CASE (RADM2,RACM_KPP,RACM_MIM_KPP, SAPRC99_KPP)
           do j=jts,jte
           do i=its,ite
            do k=kts,kte
@@ -448,6 +447,53 @@ CONTAINS
           enddo
           enddo
         endif
+!++lm New biomass burning emission package for SAPRC VBS2 mechanisms
+        CASE (SAPRC99_MOSAIC_4BIN_VBS2_KPP, SAPRC99_MOSAIC_8BIN_VBS2_AQ_KPP, SAPRC99_MOSAIC_8BIN_VBS2_KPP)
+        if( biomass_burn_opt == BIOMASSB_SAPRC ) then
+          call wrf_debug(15,'adding fire emissions for SAPRC')
+          do j=jts,jte
+            do k=kts,kte
+              do i=its,ite
+                conv_rho =(r_q*4.828e-4*dtstep)/(rho_phy(i,k,j)*60.*dz8w(i,k,j))
+                chem(i,k,j,p_co) = chem(i,k,j,p_co) + ebu(i,k,j,p_ebu_co)*conv_rho
+                chem(i,k,j,p_no) = chem(i,k,j,p_no) + ebu(i,k,j,p_ebu_no)*conv_rho
+                chem(i,k,j,p_no2) = chem(i,k,j,p_no2) + ebu(i,k,j,p_ebu_no2)*conv_rho
+                chem(i,k,j,p_so2) = chem(i,k,j,p_so2) + ebu(i,k,j,p_ebu_so2)*conv_rho
+                chem(i,k,j,p_nh3) = chem(i,k,j,p_nh3) + ebu(i,k,j,p_ebu_nh3)*conv_rho
+                chem(i,k,j,p_ch4) = chem(i,k,j,p_ch4) + ebu(i,k,j,p_ebu_ch4)*conv_rho
+                chem(i,k,j,p_acet) = chem(i,k,j,p_acet) + ebu(i,k,j,p_ebu_acet)*conv_rho
+                chem(i,k,j,p_c2h6) = chem(i,k,j,p_c2h6) + ebu(i,k,j,p_ebu_c2h6)*conv_rho
+                chem(i,k,j,p_c3h8) = chem(i,k,j,p_c3h8) + ebu(i,k,j,p_ebu_c3h8)*conv_rho
+                chem(i,k,j,p_alk3) = chem(i,k,j,p_alk3) + ebu(i,k,j,p_ebu_alk3)*conv_rho
+                chem(i,k,j,p_alk4) = chem(i,k,j,p_alk4) + ebu(i,k,j,p_ebu_alk4)*conv_rho
+                chem(i,k,j,p_alk5) = chem(i,k,j,p_alk5) + ebu(i,k,j,p_ebu_alk5)*conv_rho
+                chem(i,k,j,p_aro1) = chem(i,k,j,p_aro1) + ebu(i,k,j,p_ebu_aro1)*conv_rho
+                chem(i,k,j,p_aro2) = chem(i,k,j,p_aro2) + ebu(i,k,j,p_ebu_aro2)*conv_rho
+                chem(i,k,j,p_bald) = chem(i,k,j,p_bald) + ebu(i,k,j,p_ebu_bald)*conv_rho
+                chem(i,k,j,p_ccho) = chem(i,k,j,p_ccho) + ebu(i,k,j,p_ebu_ccho)*conv_rho
+                chem(i,k,j,p_cco_oh) = chem(i,k,j,p_cco_oh) + ebu(i,k,j,p_ebu_cco_oh)*conv_rho
+                chem(i,k,j,p_ethene) = chem(i,k,j,p_ethene) + ebu(i,k,j,p_ebu_ethene)*conv_rho
+                chem(i,k,j,p_hcho) = chem(i,k,j,p_hcho) + ebu(i,k,j,p_ebu_hcho)*conv_rho
+                chem(i,k,j,p_hcooh) = chem(i,k,j,p_hcooh) + ebu(i,k,j,p_ebu_hcooh)*conv_rho
+                chem(i,k,j,p_hono) = chem(i,k,j,p_hono) + ebu(i,k,j,p_ebu_hono)*conv_rho
+                chem(i,k,j,p_isoprene) = chem(i,k,j,p_isoprene) + ebu(i,k,j,p_ebu_isoprene)*conv_rho
+                chem(i,k,j,p_mek) = chem(i,k,j,p_mek) + ebu(i,k,j,p_ebu_mek)*conv_rho
+                chem(i,k,j,p_meoh) = chem(i,k,j,p_meoh) + ebu(i,k,j,p_ebu_meoh)*conv_rho
+                chem(i,k,j,p_methacro) = chem(i,k,j,p_methacro) + ebu(i,k,j,p_ebu_methacro)*conv_rho
+                chem(i,k,j,p_mgly) = chem(i,k,j,p_mgly) + ebu(i,k,j,p_ebu_mgly)*conv_rho
+                chem(i,k,j,p_mvk) = chem(i,k,j,p_mvk) + ebu(i,k,j,p_ebu_mvk)*conv_rho
+                chem(i,k,j,p_ole1) = chem(i,k,j,p_ole1) + ebu(i,k,j,p_ebu_ole1)*conv_rho
+                chem(i,k,j,p_ole2) = chem(i,k,j,p_ole2) + ebu(i,k,j,p_ebu_ole2)*conv_rho
+                chem(i,k,j,p_phen) = chem(i,k,j,p_phen) + ebu(i,k,j,p_ebu_phen)*conv_rho
+                chem(i,k,j,p_prod2) = chem(i,k,j,p_prod2) + ebu(i,k,j,p_ebu_prod2)*conv_rho
+                chem(i,k,j,p_rcho) = chem(i,k,j,p_rcho) + ebu(i,k,j,p_ebu_rcho)*conv_rho
+                chem(i,k,j,p_rno3) = chem(i,k,j,p_rno3) + ebu(i,k,j,p_ebu_rno3)*conv_rho
+                chem(i,k,j,p_terp) = chem(i,k,j,p_terp) + ebu(i,k,j,p_ebu_terp)*conv_rho
+              enddo
+            enddo
+          enddo
+        endif
+!--lm
     CASE DEFAULT
        call wrf_debug(15,'nothing done with burn emissions for chem array')
     END SELECT emiss_select
diff --git a/chem/module_mosaic_addemiss.F b/chem/module_mosaic_addemiss.F
index 0681ac36..8b7b875b 100644
--- a/chem/module_mosaic_addemiss.F
+++ b/chem/module_mosaic_addemiss.F
@@ -370,7 +370,7 @@ real, save :: fr8b_bburn_mosaic(8) = &
 !++ SAN 2015-04-09: Mapping fire emissions to MOSAIC 
 ! get arrays for apportioning mass into MOSAIC bins for fire emissions
         fire_inpt_select: SELECT CASE (config_flags%biomass_burn_opt)
-          CASE (BIOMASSB,BIOMASSB_MOZC)
+          CASE (BIOMASSB,BIOMASSB_MOZC,BIOMASSB_SAPRC)
             if (nsize_aer(itype) .eq. 8) then
               bburn_mosaic_f(:) = fr8b_bburn_mosaic(:)
 !++SAN
@@ -593,7 +593,7 @@ real, save :: fr8b_bburn_mosaic(8) = &
   aem_num = 0.
 
  fire_select:  SELECT CASE(config_flags%biomass_burn_opt)
-   CASE (BIOMASSB,BIOMASSB_MOZC) 
+   CASE (BIOMASSB,BIOMASSB_MOZC,BIOMASSB_SAPRC) 
       CALL wrf_debug(15,'mosaic_addemiss: adding fire emissions to MOSAIC')
 !      CALL wrf_debug(15,'e_oin_fm = PM2.5 - OC - BC')
 !      CALL wrf_debug(15,'e_oin_cm = PM10 - PM2.5')
@@ -614,7 +614,8 @@ size_loop: &
                         + bburn_mosaic_c(n)*ebu(i,k,j,p_ebu_pm10) 
 
          chem_select_3 : SELECT CASE( config_flags%chem_opt )
-            CASE(SAPRC99_MOSAIC_4BIN_VBS2_KPP) ! Set the oc to zero for VBS
+            CASE(SAPRC99_MOSAIC_4BIN_VBS2_KPP, SAPRC99_MOSAIC_8BIN_VBS2_AQ_KPP, &
+              SAPRC99_MOSAIC_8BIN_VBS2_KPP) ! Set the fire oc to zero for VBS
               aem_oc = 0.0
             END SELECT  chem_select_3 
  
diff --git a/chem/module_plumerise1.F b/chem/module_plumerise1.F
index 5e58a97c..6f1a43a4 100644
--- a/chem/module_plumerise1.F
+++ b/chem/module_plumerise1.F
@@ -199,6 +199,51 @@ subroutine plumerise_driver (id,ktau,dtstep,                           &
              enddo
            endif
          endif
+!++lm  New biomass burning emission package for SAPRC
+       elseif ( config_flags%biomass_burn_opt == BIOMASSB_SAPRC ) then
+         do j=jts,jte
+            do i=its,ite
+               ebu(i,kts,j,p_ebu_co) = ebu_in(i,1,j,p_ebu_in_co)
+               ebu(i,kts,j,p_ebu_no) = ebu_in(i,1,j,p_ebu_in_no)
+               ebu(i,kts,j,p_ebu_no2) = ebu_in(i,1,j,p_ebu_in_no2)
+               ebu(i,kts,j,p_ebu_so2) = ebu_in(i,1,j,p_ebu_in_so2)
+               ebu(i,kts,j,p_ebu_nh3) = ebu_in(i,1,j,p_ebu_in_nh3)
+               ebu(i,kts,j,p_ebu_ch4) = ebu_in(i,1,j,p_ebu_in_ch4)
+               ebu(i,kts,j,p_ebu_acet) = ebu_in(i,1,j,p_ebu_in_acet)
+               ebu(i,kts,j,p_ebu_c2h6) = ebu_in(i,1,j,p_ebu_in_c2h6)
+               ebu(i,kts,j,p_ebu_c3h8) = ebu_in(i,1,j,p_ebu_in_c3h8)
+               ebu(i,kts,j,p_ebu_alk3) = ebu_in(i,1,j,p_ebu_in_alk3)
+               ebu(i,kts,j,p_ebu_alk4) = ebu_in(i,1,j,p_ebu_in_alk4)
+               ebu(i,kts,j,p_ebu_alk5) = ebu_in(i,1,j,p_ebu_in_alk5)
+               ebu(i,kts,j,p_ebu_aro1) = ebu_in(i,1,j,p_ebu_in_aro1)
+               ebu(i,kts,j,p_ebu_aro2) = ebu_in(i,1,j,p_ebu_in_aro2)
+               ebu(i,kts,j,p_ebu_bald) = ebu_in(i,1,j,p_ebu_in_bald)
+               ebu(i,kts,j,p_ebu_ccho) = ebu_in(i,1,j,p_ebu_in_ccho)
+               ebu(i,kts,j,p_ebu_cco_oh) = ebu_in(i,1,j,p_ebu_in_cco_oh)
+               ebu(i,kts,j,p_ebu_ethene) = ebu_in(i,1,j,p_ebu_in_ethene)
+               ebu(i,kts,j,p_ebu_hcho) = ebu_in(i,1,j,p_ebu_in_hcho)
+               ebu(i,kts,j,p_ebu_hcooh) = ebu_in(i,1,j,p_ebu_in_hcooh)
+               ebu(i,kts,j,p_ebu_hono) = ebu_in(i,1,j,p_ebu_in_hono)
+               ebu(i,kts,j,p_ebu_isoprene) = ebu_in(i,1,j,p_ebu_in_isoprene)
+               ebu(i,kts,j,p_ebu_mek) = ebu_in(i,1,j,p_ebu_in_mek)
+               ebu(i,kts,j,p_ebu_meoh) = ebu_in(i,1,j,p_ebu_in_meoh)
+               ebu(i,kts,j,p_ebu_methacro) = ebu_in(i,1,j,p_ebu_in_methacro)
+               ebu(i,kts,j,p_ebu_mgly) = ebu_in(i,1,j,p_ebu_in_mgly)
+               ebu(i,kts,j,p_ebu_mvk) = ebu_in(i,1,j,p_ebu_in_mvk)
+               ebu(i,kts,j,p_ebu_ole1) = ebu_in(i,1,j,p_ebu_in_ole1)
+               ebu(i,kts,j,p_ebu_ole2) = ebu_in(i,1,j,p_ebu_in_ole2)
+               ebu(i,kts,j,p_ebu_phen) = ebu_in(i,1,j,p_ebu_in_phen)
+               ebu(i,kts,j,p_ebu_prod2) = ebu_in(i,1,j,p_ebu_in_prod2)
+               ebu(i,kts,j,p_ebu_rcho) = ebu_in(i,1,j,p_ebu_in_rcho)
+               ebu(i,kts,j,p_ebu_rno3) = ebu_in(i,1,j,p_ebu_in_rno3)
+               ebu(i,kts,j,p_ebu_terp) = ebu_in(i,1,j,p_ebu_in_terp)
+               ebu(i,kts,j,p_ebu_pm10) = ebu_in(i,1,j,p_ebu_in_pm10)
+               ebu(i,kts,j,p_ebu_pm25) = ebu_in(i,1,j,p_ebu_in_pm25)
+               ebu(i,kts,j,p_ebu_oc) = ebu_in(i,1,j,p_ebu_in_oc)
+               ebu(i,kts,j,p_ebu_bc) = ebu_in(i,1,j,p_ebu_in_bc)
+            enddo
+         enddo
+!--lm
        elseif ( config_flags%biomass_burn_opt == BIOMASSB_GHG ) then
          do j=jts,jte
             do i=its,ite
-- 
2.31.1

